plugins/resource_monitor_plugin/include/eosio/resource_monitor_plugin/file_space_handler.hpp

```diff
@@ -131,26 +133,60 @@ namespace eosio::resource_monitor {
       }
 
    void space_monitor_loop() {
+      space_monitor_task_running = true;
+
+      if (stopping) {
+         space_monitor_task_running = false;
+         return;
+      }
       if ( is_threshold_exceeded() && shutdown_on_exceeded ) {
          elog("Gracefully shutting down, exceeded file system configured threshold.");
          appbase::app().quit(); // This will gracefully stop Nodeos
+         space_monitor_task_running = false;
          return;
       }
       update_warning_interval_counter();
 
       timer.expires_from_now( boost::posix_time::seconds( sleep_time_in_secs ));
-
+      if (stopping) {
+         space_monitor_task_running = false;
+         return;
+      }
       timer.async_wait([this](auto& ec) {
          if ( ec ) {
             wlog("Exit due to error: ${ec}, message: ${message}",
                  ("ec", ec.value())
                  ("message", ec.message()));
+            space_monitor_task_running = false;
             return;
          } else {
             // Loop over
             space_monitor_loop();
          }
       });
+      space_monitor_task_running = false;
+   }
+
+   // called on main thread from plugin shutdown()
+   void stop() {
+      // signalling space_monitor_loop that plugin is being stopped
+      stopping = true;
+
+      // cancel outstanding scheduled space_monitor_loop task
+      timer.cancel();
```

Comment:
This is not thread safe as-is. `timer.cancel()` is not thread safe. You need a mutex to here to access `timer`.

Reply:
Was trying to avoid locking. Will change.